design/single-pod-helm-chart.md

```diff
@@ -0,0 +1,146 @@
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Single-pod Helm chart
+
+## Status
+
+provisional
+
+## Table of Contents
+
+<!--ts-->
+   * [Title](#title)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories [optional]](#user-stories-optional)
+            * [Story 1](#story-1)
+            * [Story 2](#story-2)
+         * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives [optional]](#alternatives-optional)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-15T11:41-05:00 -->
+
+<!--te-->
+
+## Summary
+
+Provide a Helm chart to deploy Metal3 and Ironic as a single
+pod in a Kubernetes cluster.
+
+## Motivation
+
+### Goals
+
+The goal is to support a popular way to deploy Kubernetes applications
+to simplify creation of development environments on top of arbitrary
+Kubernetes clusters.
+
+Another goal to preare to set a standard for production-grade deployment
+of Metal3 and its components.
+
+### Non-Goals
+
+Providing end-to-end bootstrap sequence for Metal3 and Ironic is not a
+a goal of this design.
+
+## Proposal
+
+### User Stories [optional]
+
+#### Story 1
+
+As a user of Metal3, I want to install it in my existing Kubernetes
+cluster using Helm.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Initial implementation includes 2 Helm charts that create corresponding
+pods in a Kubernetes cluster:
+
+* Ironic pod
+* Metal3 pod
+
+The charts shall be added as a separate repository in metal3-io space.
+
+### Risks and Mitigations
+
+None.
+
+## Design Details
+
+Helm charts will require a separate repository (metal3-helm-charts)
+to be created.
+
+In future, a CI environment that will build and test the charts,
+will have to be created.
+
+Potentially, the Helm charts will require changes in the way the
+Docker images for components of Ironic and Metal3 are built. The
+changes will include additional parameters that will be exposed
+through the charts metadata (values.yaml files).
+
+### Work Items
+
+1. Create a Helm chart for Ironic and its components:
+   * ``ironic``
+   * ``ironic-inspector``
+   * ``ironic-dnsmasq``
+   * ``ironic-httpd``
+   * ``mariadb``
+2. Create a Helm chart for Metal3 components:
+   * ``baremetal-operator``.
```

Comment:
Minor thing, right now I think we expect Ironic to be in the same pod as baremetal-operator

Reply:
Thank you, Russel. This is a part of our attempt to evaluate scaling strategies for the future. Probably such separation, if needed at all, will make more sense a few iterations down the road.

Reply:
I agree with @russellb that the initial target should a single self-contained pod.  Scaling will also require looking at how well the operator itself deals with scaling (just in terms of being able to Reconcile a very large number of hosts) before determining the right architecture is splitting out and independently scaling ironic versus (just as an example) sharding across self-contained pods.

Reply:
Given that we will need multiple instances of Ironic to handle multiple group of nodes (for example, multiple racks), baremetal-operator will need to handle groups of BareMetalHost objects based on labels or something else. So it could be either one instance of baremetal-operator per group, or single instance managing all groups through multiple Ironic API endpoints. I have no immediate answer what it should be, and it seems to be part of another discussion anyways.

Self-contained pod should the current goal anyways. I will reflect this in the design.

Reply:
and by all means feel free to experiment with alternative deployment models.  I don't think anything is particularly enforced, except baremetal-operator assumes the ironic API is reachable via localhost at the moment.  That's easily fixable, but if it's remote we'll have to put some proper security in place.

All of this aside, I think the important thing with this whole PR is "should the metal3-io project should host a repo where people collaborate on helm charts for metal3-io" and IMO, it's yes